src/GPS.cpp

```diff
@@ -0,0 +1,110 @@
+using namespace marnav;
+using namespace std;
+using namespace gps_base;
+
+GPS_SOLUTION_TYPES GPS::getPositionType(utils::optional<nmea::mode_indicator> mode)
+{
+    switch (*mode) {
+        case nmea::mode_indicator::invalid:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::autonomous:
+            return GPS_SOLUTION_TYPES::AUTONOMOUS;
+        case nmea::mode_indicator::differential:
+            return GPS_SOLUTION_TYPES::DIFFERENTIAL;
+        case nmea::mode_indicator::estimated:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::manual_input:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::simulated:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::data_not_valid:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::precise:
+            return GPS_SOLUTION_TYPES::AUTONOMOUS;
+        default:
+            return GPS_SOLUTION_TYPES::INVALID;
+    }
+}
+
+base::Time GPS::buildRockTime(utils::optional<nmea::time> const& optional_time,
+    utils::optional<nmea::date> const& optional_date)
+{
+    auto date = optional_date.value();
+    auto time = optional_time.value();
+    return base::Time::fromTimeValues(date.year() + 2000,
+        static_cast<int>(date.mon()),
+        date.day(),
+        time.hour(),
+        time.minutes(),
+        time.seconds(),
+        time.milliseconds(),
+        0);
+}
+
+Solution GPS::getSolution(nmea::rmc const& rmc, nmea::gsa const& gsa)
+{
+    Solution solution;
+    auto mode_indicator = rmc.get_mode_ind().value();
+    solution.positionType = getPositionType(mode_indicator);
+    if (solution.positionType != GPS_SOLUTION_TYPES::INVALID) {
+        solution.time = buildRockTime(rmc.get_time_utc(), rmc.get_date());
+        auto optional_latitude = rmc.get_latitude();
+        auto optional_longitude = rmc.get_longitude();
+        if (optional_latitude.has_value()) {
+            solution.latitude = optional_latitude.value();
+        }
+        if (optional_longitude.has_value()) {
+            solution.longitude = optional_longitude.value();
+        }
+    }
+    else {
+        solution.time = base::Time::now();
+        solution.latitude = base::unknown<double>();
+        solution.longitude = base::unknown<double>();
+    }
+    solution.noOfSatellites = 0;
+    for (int i = 0; i < gsa.max_satellite_ids; i++) {
+        if (gsa.get_satellite_id(i).has_value()) {
+            solution.noOfSatellites += 1;
+        }
+    }
+    return solution;
+}
+
+SolutionQuality GPS::getSolutionQuality(nmea::gsa const& gsa)
+{
+    SolutionQuality solution_quality;
+    solution_quality.time = base::Time::now();
+    auto optional_pdop = gsa.get_pdop();
+    if (optional_pdop.has_value()) {
+        solution_quality.pdop = optional_pdop.value();
+    }
+    else {
+        solution_quality.pdop = base::unknown<double>();
+    }
```

Comment:
We can use cpp17, so you could do:
```
auto pdop = gsa.get_pdop();
solution_quality.pdop = pdop.value_or(base::unknown<double>());
```

Same for the rest of the optionals you use

https://en.cppreference.com/w/cpp/utility/optional/value_or

Reply:
Realised that we are using the optional from marnav, still I am going to leave this comment as a FYI for when you work with c++ optionals